### PR TITLE
feat: add interface function for updating learning_rate per each iteration in LightGBMDelegate

### PR DESCRIPTION
**Pull request by ocworld**
Interface function is added for updating learning_rate per each iteration in LightGBMDelegate.

It is need if researcher want to decreasing or increasing learning rate per each boost round.

### CODE REVIEW

**Comment by ocworld** [comment]
@AhnLab-OSS

**Comment by ocworld** [comment]
@imatiach-msft  Can I see detailed failure logs in unittest?

**Comment by imatiach-msft** [comment]
Job aborted due to stage failure: Task not serializable: java.io.NotSerializableException: org.scalatest.Assertions$AssertionsHelper Serialization stack: - object not serializable (class: org.scalatest.Assertions$AssertionsHelper, value: org.scalatest.Assertions$AssertionsHelper@15d041ed) - field (class: org.scalatest.FunSuite, name: assertionsHelper, type: class org.scalatest.Assertions$AssertionsHelper) - object (class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier, VerifyLightGBMClassifier) - field (class: com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21, name: $outer, type: class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier) - object (class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21, <function0>) - field (class: com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1, name: $outer, type: class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21) - object (class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1, com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1@7e609acb) - field (class: scala.Some, name: x, type: class java.lang.Object) - object (class scala.Some, Some(com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1@7e609acb)) - field (class: com.microsoft.ml.spark.lightgbm.LightGBMClassifier, name: delegate, type: class scala.Option) - object (class com.microsoft.ml.spark.lightgbm.LightGBMClassifier, LightGBMClassifier_58b8194d6ebe) - field (class: com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6, name: $outer, type: interface com.microsoft.ml.spark.lightgbm.LightGBMBase) - object (class com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6, <function1>) - field (class: org.apache.spark.sql.execution.MapPartitionsExec, name: func, type: interface scala.Function1) - object (class org.apache.spark.sql.execution.MapPartitionsExec, MapPartitions <function1>, obj#29842: com.microsoft.ml.spark.lightgbm.LightGBMBooster +- DeserializeToObject createexternalrow(labels#29720, newInstance(class org.apache.spark.ml.linalg.VectorUDT).deserialize, StructField(labels,DoubleType,false), StructField(features,org.apache.spark.ml.linalg.VectorUDT@3bfc3ba7,true)), obj#29841: org.apache.spark.sql.Row +- Coalesce 2 +- *(1) Project [labels#28546 AS labels#29720, features#29074] +- *(1) Sample 0.0, 0.8, false, 42 +- *(1) Project [labels#28546, features#29074] +- *(1) Sort [age#27948 ASC NULLS FIRST, balance#27953 ASC NULLS FIRST, day#27957 ASC NULLS FIRST, month#27958 ASC NULLS FIRST, duration#27959 ASC NULLS FIRST, campaign#27960 ASC NULLS FIRST, pdays#27961 ASC NULLS FIRST, previous#27962 ASC NULLS FIRST, poutcome#27963 ASC NULLS FIRST, c_job#28349 ASC NULLS FIRST, c_marital#28368 ASC NULLS FIRST, c_education#28388 ASC NULLS FIRST, c_default#28409 ASC NULLS FIRST, c_housing#28431 ASC NULLS FIRST, c_loan#28454 ASC NULLS FIRST, c_contact#28478 ASC NULLS FIRST, labels#28546 ASC NULLS FIRST, features#29074 ASC NULLS FIRST], false, 0 +- InMemoryTableScan [age#27948, balance#27953, day#27957, month#27958, duration#27959, campaign#27960, pdays#27961, previous#27962, poutcome#27963, c_job#28349, c_marital#28368, c_education#28388, c_default#28409, c_housing#28431, c_loan#28454, c_contact#28478, labels#28546, features#29074] +- InMemoryRelation [age#27948, balance#27953, day#27957, month#27958, duration#27959, campaign#27960, pdays#27961, previous#27962, poutcome#27963, c_job#28349, c_marital#28368, c_education#28388, c_default#28409, c_housing#28431, c_loan#28454, c_contact#28478, labels#28546, features#29074], StorageLevel(disk, memory, deserialized, 1 replicas) +- *(1) Project [age#27948, balance#27953, day#27957, month#27958, durati



org.apache.spark.SparkException: Job aborted due to stage failure: Task not serializable: java.io.NotSerializableException: org.scalatest.Assertions$AssertionsHelper
Serialization stack:
	- object not serializable (class: org.scalatest.Assertions$AssertionsHelper, value: org.scalatest.Assertions$AssertionsHelper@15d041ed)
	- field (class: org.scalatest.FunSuite, name: assertionsHelper, type: class org.scalatest.Assertions$AssertionsHelper)
	- object (class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier, VerifyLightGBMClassifier)
	- field (class: com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21, name: $outer, type: class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier)
	- object (class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21, <function0>)
	- field (class: com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1, name: $outer, type: class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21)
	- object (class com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1, com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1@7e609acb)
	- field (class: scala.Some, name: x, type: class java.lang.Object)
	- object (class scala.Some, Some(com.microsoft.ml.spark.lightgbm.split1.VerifyLightGBMClassifier$$anonfun$21$TrainDelegate$1@7e609acb))
	- field (class: com.microsoft.ml.spark.lightgbm.LightGBMClassifier, name: delegate, type: class scala.Option)
	- object (class com.microsoft.ml.spark.lightgbm.LightGBMClassifier, LightGBMClassifier_58b8194d6ebe)
	- field (class: com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6, name: $outer, type: interface com.microsoft.ml.spark.lightgbm.LightGBMBase)
	- object (class com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6, <function1>)
	- field (class: org.apache.spark.sql.execution.MapPartitionsExec, name: func, type: interface scala.Function1)
	- object (class org.apache.spark.sql.execution.MapPartitionsExec, MapPartitions <function1>, obj#29842: com.microsoft.ml.spark.lightgbm.LightGBMBooster
+- DeserializeToObject createexternalrow(labels#29720, newInstance(class org.apache.spark.ml.linalg.VectorUDT).deserialize, StructField(labels,DoubleType,false), StructField(features,org.apache.spark.ml.linalg.VectorUDT@3bfc3ba7,true)), obj#29841: org.apache.spark.sql.Row
   +- Coalesce 2
      +- *(1) Project [labels#28546 AS labels#29720, features#29074]
         +- *(1) Sample 0.0, 0.8, false, 42
            +- *(1) Project [labels#28546, features#29074]
               +- *(1) Sort [age#27948 ASC NULLS FIRST, balance#27953 ASC NULLS FIRST, day#27957 ASC NULLS FIRST, month#27958 ASC NULLS FIRST, duration#27959 ASC NULLS FIRST, campaign#27960 ASC NULLS FIRST, pdays#27961 ASC NULLS FIRST, previous#27962 ASC NULLS FIRST, poutcome#27963 ASC NULLS FIRST, c_job#28349 ASC NULLS FIRST, c_marital#28368 ASC NULLS FIRST, c_education#28388 ASC NULLS FIRST, c_default#28409 ASC NULLS FIRST, c_housing#28431 ASC NULLS FIRST, c_loan#28454 ASC NULLS FIRST, c_contact#28478 ASC NULLS FIRST, labels#28546 ASC NULLS FIRST, features#29074 ASC NULLS FIRST], false, 0
                  +- InMemoryTableScan [age#27948, balance#27953, day#27957, month#27958, duration#27959, campaign#27960, pdays#27961, previous#27962, poutcome#27963, c_job#28349, c_marital#28368, c_education#28388, c_default#28409, c_housing#28431, c_loan#28454, c_contact#28478, labels#28546, features#29074]
                        +- InMemoryRelation [age#27948, balance#27953, day#27957, month#27958, duration#27959, campaign#27960, pdays#27961, previous#27962, poutcome#27963, c_job#28349, c_marital#28368, c_education#28388, c_default#28409, c_housing#28431, c_loan#28454, c_contact#28478, labels#28546, features#29074], StorageLevel(disk, memory, deserialized, 1 replicas)
                              +- *(1) Project [age#27948, balance#2795

**Comment by imatiach-msft** [comment]
[info] - Verify LightGBM Classifier updating learning_rate on training by using LightGBMDelegate *** FAILED ***

[info] 	- object (class scala.collection.immutable.List$SerializationProxy, scala.collection.immutable.List$SerializationProxy@3d4ebb56)

[info] 	- writeReplace data (class: scala.collection.immutable.List$SerializationProxy)

[info] 	- object (class scala.collection.immutable.$colon$colon, List(org.apache.spark.OneToOneDependency@230f8966))

[info] 	- field (class: org.apache.spark.rdd.RDD, name: org$apache$spark$rdd$RDD$$dependencies_, type: interface scala.collection.Seq)

[info] 	- object (class org.apache.spark.rdd.MapPartitionsRDD, MapPartitionsRDD[4766] at reduce at LightGBMBase.scala:160)

[info] 	- field (class: scala.Tuple2, name: _1, type: class java.lang.Object)

[info] 	- object (class scala.Tuple2, (MapPartitionsRDD[4766] at reduce at LightGBMBase.scala:160,<function2>))

[info]   at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:40)

[info]   at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:46)

[info]   at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)

[info]   at org.apache.spark.scheduler.DAGScheduler.submitMissingTasks(DAGScheduler.scala:1155)

[info]   at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$submitStage(DAGScheduler.scala:1069)

[info]   at org.apache.spark.scheduler.DAGScheduler.handleJobSubmitted(DAGScheduler.scala:1013)

[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2067)

[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)

[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)

[info]   at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)

[info]   ...

**Comment by imatiach-msft** [comment]
@ocworld  it looks like there was a serialization failure in the test case, will take a look and try to run locally when I get a chance

File: src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala [hunk 1]
```diff
@@ -201,10 +201,19 @@ private object TrainUtils extends Serializable {
...
     val delegate = trainParams.delegate
     val partitionId = TaskContext.getPartitionId
+    var learningRate: Double = trainParams.learningRate
     while (!isFinished && iters < trainParams.numIterations) {
 
       if (delegate.isDefined) {
         delegate.get.beforeTrainIteration(partitionId, iters, log, trainParams, boosterPtr, hasValid)
+        val newLearningRate = delegate.get.getLearningRate(partitionId, iters, log, trainParams, learningRate)
+        if (newLearningRate != learningRate) {
+          log.info(s"LightGBM worker calling LGBM_BoosterResetParameter to reset learningRate" +
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/TrainUtils.scala:211]: nice logging!

File: src/test/scala/com/microsoft/ml/spark/lightgbm/split1/VerifyLightGBMClassifier.scala [hunk 2]
```diff
@@ -360,6 +362,47 @@ class VerifyLightGBMClassifier extends Benchmarks with EstimatorFuzzing[LightGBM
     assert(metric > 0.8)
   }
 
+  test("Verify LightGBM Classifier updating learning_rate on training by using LightGBMDelegate") {
+
+    class TrainDelegate extends LightGBMDelegate {
```
Comment (imatiach-msft) [inline src/test/scala/com/microsoft/ml/spark/lightgbm/split1/VerifyLightGBMClassifier.scala:367]: maybe the test is failing because this class is not serializable?

**Comment by imatiach-msft** [comment]
maybe add this to the class:

https://alvinalexander.com/scala/how-to-use-serialization-in-scala-serializable-trait/

give it a @SerialVersionUID(100L) and "extends Serializable"

**Comment by ocworld** [comment]
@imatiach-msft Thanks to your comments and logs. 
LightGBMDelegate already extend serializable.

So, Two things are fixed.
1) @serialversionuid(100L), you mentioned, is written in my code.
2) In unittest, TrainDelegate Class is inner class. I assume that outter class, VerifyLightGBMClassifier, is not serializable. When I tested on my local sample program, this case cause not serialization error.
So, TrainDelegate is changed to independent class.

**Comment by ocworld** [comment]
@imatiach-msft Can I see error logs?

**Comment by imatiach-msft** [comment]
@ocworld it looks like some new test infrastructure issue, let me see if @mhamilton723 might take a look since he is more familiar with the cognitive tests
